### DISCUSSION TITLE
How I implement "Spatie/laravel-permission" in a completed project

### DISCUSSION THREAD

**Discussion by M-Valentina-H** [spatie/laravel-permission#1805] [Q&A] [answered]
How would I implement the `spatie/laravel-permission` in a project that I already have almost complete? Let's say that it is almost finished, I only need to define the permissions of the roles.
I already have the role, permission and user tables, their relationships and their columns defined (with another name for my misfortune), I don't need to run the library migration. I know that you can change the name of the tables in the section `table_names' =>` of the file `config/permission.php` but when I try to do something from the library it asks me for columns like `role_id` (which I do not have because in my case it is called `rol_id`) and it gives me an error.
Is there no way to modify the library so that it changes the columns it calls as in the case of the table names in the `permission.php` file?

**Comment by stychos** [comment]
Take a look at `src/Models/Permission.php`, it can be easily modified to use custom columns in relations.

**Reply by M-Valentina-H** [reply]
I already modified the whole file and it didn't work. It's like I haven't done anything. But thanks for answering

**Comment by erikn69** [comment, marked as answer]
On the config file you only can change the table names and user foreign key for custom ones, fields like `name` and `guard_name` can't change, 
If  you want to change `role_id` and `permission_id` pivots you have to create models `Role` and `Permission` extends from spatie models, and override relations and replace default foreign keys 
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Models/Role.php#L48-L70
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Models/Permission.php#L51-L73
Also, you have to override the traits, (On User model)
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Traits/HasRoles.php#L40-L49
https://github.com/spatie/laravel-permission/blob/a6e4122b65094baba7f98df153af0768ef910c85/src/Traits/HasPermissions.php#L43-L52
And now must be working, a lot overrides, maybe can be a new feature, you have to ask to @freekmurze @drbyte, looks easy to set a config for this , i have time for make a PR

**Reply by M-Valentina-H** [reply]
Thank you very much for taking the time to write this answer, I'm sorry, I had not understood anything, but yesterday when I was with this all day debugging and today too, I noticed some things that made me make sense of your answer, I just understood LOL. Right now I am modifying my models `Rol` (Role) and `Permiso` (Permission) and adapting them to be like the models that are in `laravel-permission/src/Models/`,  for now it works, I will modify the migration and thanks to you, I know that I must also modify the files that are in `laravel-permission/src/Traits/`.
Let's see how it turns out, again, thank you very much... <3

P.S: I apologize if I am not well understood, I am using google translator

**Reply by erikn69** [reply]
I am using google translator too (fail a lot but helps, y talvez pude escribir en otro idioma pero debe quedar la respuesta para todos), 

> On User model

Only override the relation `roles()` and  `permissions()` on your models, (the one with `use hasRoles;`  normaly `\App\Models\User.php`)

I can add support for this but i don`t have any answer from package collaborators

**Reply by erikn69** [reply]
Team's support and custom pivot names added
